### PR TITLE
Introduce scope to semantic analyzer and update correlation in Apply node (part 1)

### PR DESCRIPTION
**Pull request by kokosing**
Introduce scope to semantic analyzer and update correlation in Apply node

### CODE REVIEW

**Comment by martint** [comment]
I merged all commits except for "Use SubqueryPlanner in RelationPlanner". I'm reviewing that now.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 1]
```diff
@@ -276,14 +277,24 @@ public void recordSubqueries(Node node, ExpressionAnalysis expressionAnalysis)
...
     }
 
-    public JoinInPredicates getJoinInPredicates(Join node)
+    public JoinSideExpressions getJoinSideExpressions(Join node)
     {
-        return joinInPredicates.get(node);
+        return joinSideExpressions.get(node);
+    }
+
+    public void setPostJoinConjunct(Join node, Expression postJoinConjunct)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: The word "conjunct" is only relevant when describing terms of an AND expression. If anything, this should be called "postJoinPredicate".

Speaking of which, how is this different from the a regular "where" predicate? Any predicates that are applied after the join are semantically equivalent to being in the WHERE clause.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 2]
```diff
@@ -435,31 +446,31 @@ public double getSampleRatio(SampledRelation relation)
         return sampleRatios.get(relation);
     }
 
-    public static class JoinInPredicates
+    public static class JoinSideExpressions
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: What does "join side" refer to?

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: Is it "join side" expressions? Or join "side expressions"? I'm trying to understand what this class stands for.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 3]
```diff
@@ -270,12 +268,12 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
                 }
             }
 
-            Analysis.JoinInPredicates joinInPredicates = analysis.getJoinInPredicates(node);
+            Analysis.JoinSideExpressions joinSideExpressions = analysis.getJoinSideExpressions(node);
 
             // Add semi joins if necessary
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Is this comment still accurate? It could be any subquery now, no?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 4]
```diff
@@ -1977,38 +1977,97 @@ public void testJoinWithMultipleInSubqueryClauses()
                 queryPrefix + "(x in (VALUES 1,2)) = (y in (VALUES 1,2)) AND (x in (VALUES 1)) = (y in (VALUES 3))",
                 "VALUES (2,2), (2,1)");
         assertQuery(
-                queryPrefix + "(x in (VALUES 1,2)) = (y in (VALUES 1,2)) AND (x in (VALUES 1)) != (y in (VALUES 3))",
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: What's the purpose of this change?

Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Sorry. It was a mistake. Restoring previous version.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: Exactly, `postJoindPredicate` is a WHERE predicate applied just after the join.

When I was applying your comments, I realized that this `postJoinPredicate` is actually known to `RelationPlanner`. `RelationPlanner` retrieves that from regular `joinCriteria`. The same thing for JoinSideExpressions. So I removed them from `Analysis`.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 2 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: It was "join side" expressions. But I have just removed this entity, see my previous comment. These expression were already known to `RelationPlanner` so there is no need to store it in `Analysis`.

**Comment by kokosing** [comment]
@martint All the comments applied, rebased.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java [hunk 5]
```diff
@@ -111,12 +117,53 @@ public void testPlanDsl()
...
     }
 
+    @Test
+    public void testSameScalarSubqueryIsAppliedOnlyOnce()
+    {
+        // three subqueries with two duplicates, only two scalar joins should be in plan
+        Plan plan = plan("SELECT * FROM orders WHERE orderkey = (SELECT 1) AND custkey = (SELECT 2) AND custkey != (SELECT 1)");
+        PlanNodeExtractor planNodeExtractor = new PlanNodeExtractor(planNode -> planNode instanceof EnforceSingleRowNode);
+        plan.getRoot().accept(planNodeExtractor, null);
+        assertTrue(planNodeExtractor.getNodes().size() == 2);
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java]: Use assertEquals

**Comment by martint** [comment]
Merged, thanks!
